File: apps/web/src/services/api/crawlSources.ts
/**
 * Crawl sources API client (FR-002-2). GET/POST/PATCH/DELETE /api/v1/crawl-sources.
 */

const BASE = import.meta.env.VITE_API_BASE_URL ?? '/api/v1'

export interface CrawlSourceItem {
  id: number
  base_url: string
  source_name: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface CrawlSourceListResponse {
  items: CrawlSourceItem[]
  total: number
}

export interface CrawlSourceCreate {
  base_url: string
  source_name: string
  is_active?: boolean
}

export interface CrawlSourceUpdate {
  base_url?: string
  source_name?: string
  is_active?: boolean
}

export async function getCrawlSources (params?: {
  is_active?: boolean
  limit?: number
  offset?: number
}): Promise<CrawlSourceListResponse> {
  const search = new URLSearchParams()
  if (params?.is_active != null) search.set('is_active', String(params.is_active))
  if (params?.limit != null) search.set('limit', String(params.limit))
  if (params?.offset != null) search.set('offset', String(params.offset))
  const qs = search.toString()
  const url = `${BASE}/crawl-sources${qs ? `?${qs}` : ''}`
  const res = await fetch(url)
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

export async function getCrawlSource (id: number): Promise<CrawlSourceItem> {
  const res = await fetch(`${BASE}/crawl-sources/${id}`)
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

export async function createCrawlSource (body: CrawlSourceCreate): Promise<CrawlSourceItem> {
  const res = await fetch(`${BASE}/crawl-sources`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

export async function updateCrawlSource (id: number, body: CrawlSourceUpdate): Promise<CrawlSourceItem> {
  const res = await fetch(`${BASE}/crawl-sources/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
  return res.json()
}

export async function deleteCrawlSource (id: number): Promise<void> {
  const res = await fetch(`${BASE}/crawl-sources/${id}`, { method: 'DELETE' })
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`)
}

File: apps/api/src/api/crawl_sources.py
"""Crawl sources API: GET/POST/PATCH/DELETE /crawl-sources (FR-002-2)."""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_async_session
from services.crawl_source_service import CrawlSourceService

router = APIRouter()


class CrawlSourceItem(BaseModel):
    id: int
    base_url: str
    source_name: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CrawlSourceListResponse(BaseModel):
    items: list[CrawlSourceItem]
    total: int


class CrawlSourceCreate(BaseModel):
    base_url: str
    source_name: str
    is_active: bool = True


class CrawlSourceUpdate(BaseModel):
    base_url: str | None = None
    source_name: str | None = None
    is_active: bool | None = None


@router.get("/crawl-sources", response_model=CrawlSourceListResponse)
async def list_crawl_sources(
    is_active: bool | None = Query(None, description="Filter by active"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    session: AsyncSession = Depends(get_async_session),
):
    svc = CrawlSourceService(session)
    items, total = await svc.list(is_active=is_active, limit=limit, offset=offset)
    return CrawlSourceListResponse(items=[CrawlSourceItem.model_validate(x) for x in items], total=total)


@router.get("/crawl-sources/{id}", response_model=CrawlSourceItem)
async def get_crawl_source(
    id: int,
    session: AsyncSession = Depends(get_async_session),
):
    svc = CrawlSourceService(session)
    row = await svc.get(id)
    if row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Crawl source not found")
    return CrawlSourceItem.model_validate(row)


@router.post("/crawl-sources", response_model=CrawlSourceItem, status_code=status.HTTP_201_CREATED)
async def create_crawl_source(
    body: CrawlSourceCreate,
    session: AsyncSession = Depends(get_async_session),
):
    svc = CrawlSourceService(session)
    row = await svc.create(
        base_url=body.base_url,
        source_name=body.source_name,
        is_active=body.is_active,
    )
    await session.commit()
    return CrawlSourceItem.model_validate(row)


@router.patch("/crawl-sources/{id}", response_model=CrawlSourceItem)
async def update_crawl_source(
    id: int,
    body: CrawlSourceUpdate,
    session: AsyncSession = Depends(get_async_session),
):
    svc = CrawlSourceService(session)
    row = await svc.update(
        id,
        base_url=body.base_url,
        source_name=body.source_name,
        is_active=body.is_active,
    )
    if row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Crawl source not found")
    await session.commit()
    return CrawlSourceItem.model_validate(row)


@router.delete("/crawl-sources/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_crawl_source(
    id: int,
    session: AsyncSession = Depends(get_async_session),
):
    svc = CrawlSourceService(session)
    ok = await svc.delete(id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Crawl source not found")
    await session.commit()
